File: app/companion.js
// Phone ⇄ TV companion remote.
//
// A phone can't reach the TV directly, so both ends talk to a relay over two
// plain HTTP verbs — SSE to listen, POST to send — addressed by an 8-character
// room code. The LAN relay (server.mjs, which Electron also hosts) and the
// hosted relay (api/companion.js on Vercel) speak the same API, so this client
// never knows which one it is on.
//
// Phone messages are untrusted input: parseCommand() is the single gate every
// one passes before the TV acts on it. Everything except openLink() is pure, so
// it unit-tests without a DOM.

// No I/O/0/1 — a code read off a TV shouldn't be ambiguous. 32 symbols, so a
// random byte masked to 5 bits picks uniformly.
const ALPHABET = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
export const ROOM_RE = /^[A-HJ-NP-Z2-9]{8}$/;
export const ROLES = ['tv', 'phone'];
// TV → phone state carries the queue (with artwork URLs), so the cap has room
// for QUEUE_MAX songs at ~300 bytes each.
export const MAX_MESSAGE_BYTES = 16384;
export const QUEUE_MAX = 30;

const NUDGES_MS = [-100, -25, 25, 100];

export function newRoomCode() {
  return [...crypto.getRandomValues(new Uint8Array(8))].map((b) => ALPHABET[b & 31]).join('');
}

function text(v, max = 200) {
  return typeof v === 'string' ? v.trim().slice(0, max) : '';
}

/**
 * Validate a song picked on the phone.
 * @returns {{artist:string, track:string, duration?:number, artwork?:string} | null}
 */
export function parseSong(s) {
  if (!s || typeof s !== 'object') return null;
  const track = text(s.track);
  if (!track) return null;
  const song = { artist: text(s.artist), track };
  if (Number.isFinite(s.duration) && s.duration > 0 && s.duration < 3600) song.duration = s.duration;
  const art = text(s.artwork, 500);
  if (art.startsWith('https://')) song.artwork = art;
  return song;
}

/**
 * Normalise one phone → TV message, or null if it isn't a valid command.
 * Unknown fields are dropped; nothing from the phone reaches the app unshaped.
 */
export function parseCommand(msg) {
  if (!msg || typeof msg !== 'object') return null;
  switch (msg.type) {
    case 'hello':
    case 'toggle':
    case 'change':
    case 'next':
      return { type: msg.type };
    case 'play':
    case 'queue': {
      const song = parseSong(msg.song);
      return song ? { type: msg.type, song } : null;
    }
    case 'unqueue':
      return Number.isInteger(msg.index) && msg.index >= 0 ? { type: 'unqueue', index: msg.index } : null;
    case 'nudge':
      return NUDGES_MS.includes(msg.ms) ? { type: 'nudge', ms: msg.ms } : null;
    case 'feel':
      return msg.sense === 'early' || msg.sense === 'late' ? { type: 'feel', sense: msg.sense } : null;
    default:
      return null;
  }
}

/**
 * Has the song on screen run out? True once the clock passes both the last
 * lyric line and the timeline's duration, plus an outro grace.
 */
export function songFinished({ now, timeline, graceSec = 6 }) {
  const lines = timeline?.lines;
  if (!lines?.length || !Number.isFinite(now)) return false;
  const end = Math.max(lines[lines.length - 1].end || 0, timeline.duration || 0);
  return end > 0 && now > end + graceSec;
}

export function relayUrl(base, room, role) {
  return `${base}/api/companion?room=${room}&role=${role}`;
}

/**
 * Open one end of a room: SSE in, POST out. The POST body is text/plain so it
 * stays a CORS "simple request" — the Electron renderer (file:// origin) needs
 * no preflight. EventSource reconnects on its own after drops.
 */
export function openLink({ base, room, role, onMessage, onStatus = () => {} }) {
  const url = relayUrl(base, room, role);
  const es = new EventSource(url);
  es.onopen = () => onStatus('open');
  es.onerror = () => onStatus('retrying');
  es.onmessage = (e) => {
    let msg;
    try {
      msg = JSON.parse(e.data);
    } catch {
      return;
    }
    onMessage(msg);
  };
  return {
    send: (msg) =>
      fetch(url, { method: 'POST', headers: { 'Content-Type': 'text/plain' }, body: JSON.stringify(msg) })
        .catch(() => {}),
    close: () => es.close(),
  };
}

File: tvos/Bar4BarKit/Sources/Bar4BarKit/Companion.swift
// Phone ⇄ TV remote, the TV end. Same protocol and hosted relay as the web app
// (app/companion.js, api/companion.js): SSE to listen, text/plain POST to send,
// addressed by an 8-character room code. The phone page is the existing
// companion.html, so a phone can't tell a native TV from the web one.
//
// Phone messages are untrusted: parseCommand is the only way one reaches the app.

import Foundation

public enum PhoneCommand: Equatable, Sendable {
  case hello(by: String?)
  case toggle, change, next
  /// The song carries the guest's name in `by`.
  case play(Song), queue(Song)
  case unqueue(Int)
  case nudge(ms: Int)
  case feel(early: Bool)
}

/// TV → phone snapshot of what's on screen.
public struct TVState: Encodable, Equatable, Sendable {
  public struct Queued: Encodable, Equatable, Sendable {
    public var track: String
    public var artist: String
    public var artwork: String?
    public var by: String?
  }

  public var type = "state"
  public var mode: String // "playing" | "setup"
  public var track: String
  public var artist: String
  public var art: String
  public var playing: Bool
  public var offsetMs: Int
  public var queue: [Queued]

  public init(mode: String, track: String, artist: String, art: String, playing: Bool, offsetMs: Int, queue: [Song]) {
    self.mode = mode
    self.track = track
    self.artist = artist
    self.art = art
    self.playing = playing
    self.offsetMs = offsetMs
    self.queue = queue.map { Queued(track: $0.track, artist: $0.artist, artwork: $0.artwork, by: $0.by) }
  }
}

public enum Companion {
  public static let base = URL(string: "https://smartlyric.vercel.app")!
  public static let queueMax = 30
  public static let nameMax = 24
  static let nudgesMs: Set<Double> = [-100, -25, 25, 100]
  // No I/O/0/1 — a code read off a TV shouldn't be ambiguous.
  static let alphabet = Array("ABCDEFGHJKLMNPQRSTUVWXYZ23456789")

  public static func newRoomCode() -> String {
    String((0..<8).map { _ in alphabet.randomElement()! }) // SystemRandomNumberGenerator: CSPRNG
  }

  public static func phoneURL(room: String, base: URL = base) -> URL {
    URL(string: "\(base.absoluteString)/companion.html?room=\(room)")!
  }

  /// The rotation: round-robin the queue by whoever added each song, the way a
  /// karaoke host runs a room — three songs from one guest don't lock everyone
  /// else out. Stable inside each singer's own list, and a no-op when every entry
  /// came from the same phone (or from the TV, where `by` is nil).
  /// Port of `fairOrder` in app/companion.js.
  public static func fairOrder(_ queue: [Song]) -> [Song] {
    var lanes: [String: [Song]] = [:]
    var order: [String] = [] // first-seen order, so the rotation is deterministic
    for song in queue {
      let who = song.by ?? ""
      if lanes[who] == nil {
        lanes[who] = []
        order.append(who)
      }
      lanes[who]?.append(song)
    }
    guard order.count > 1 else { return queue }

    var out: [Song] = []
    out.reserveCapacity(queue.count)
    var i = 0
    while out.count < queue.count {
      let who = order[i % order.count]
      if var lane = lanes[who], !lane.isEmpty {
        out.append(lane.removeFirst())
        lanes[who] = lane
      }
      i += 1
    }
    return out
  }

  static func relayURL(room: String, base: URL = base) -> URL {
    URL(string: "\(base.absoluteString)/api/companion?room=\(room)&role=tv")!
  }

  /// Display name: control characters removed, trimmed, capped. "" if none.
  public static func parseName(_ v: Any?) -> String {
    guard let s = v as? String else { return "" }
    let kept = s.unicodeScalars.filter { $0.value > 0x1f && $0.value != 0x7f }
    return String(String(String.UnicodeScalarView(kept)).trimmingCharacters(in: .whitespacesAndNewlines).prefix(nameMax))
  }

  static func text(_ v: Any?, max: Int = 200) -> String {
    String(((v as? String) ?? "").trimmingCharacters(in: .whitespacesAndNewlines).prefix(max))
  }

  /// A JSON number that isn't a boolean (NSNumber bridges both).
  static func number(_ v: Any?) -> Double? {
    guard let n = v as? NSNumber, CFGetTypeID(n) != CFBooleanGetTypeID() else { return nil }
    return n.doubleValue
  }

  public static func parseSong(_ v: Any?) -> Song? {
    guard let s = v as? [String: Any] else { return nil }
    let track = text(s["track"])
    guard !track.isEmpty else { return nil }
    var song = Song(track: track, artist: text(s["artist"]))
    if let d = number(s["duration"]), d.isFinite, d > 0, d < 3600 { song.duration = d }
    let art = text(s["artwork"], max: 500)
    if art.hasPrefix("https://") { song.artwork = art }
    return song
  }

  public static func parseCommand(_ data: Data) -> PhoneCommand? {
    guard let msg = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else { return nil }
    let by = parseName(msg["by"])
    switch msg["type"] as? String {
    case "hello":
      return .hello(by: by.isEmpty ? nil : by)
    case "toggle": return .toggle
    case "change": return .change
    case "next": return .next
    case let type? where type == "play" || type == "queue":
      guard var song = parseSong(msg["song"]) else { return nil }
      if !by.isEmpty { song.by = by }
      return type == "play" ? .play(song) : .queue(song)
    case "unqueue":
      guard let i = number(msg["index"]), i >= 0, i == i.rounded(), i < 1e9 else { return nil }
      return .unqueue(Int(i))
    case "nudge":
      guard let ms = number(msg["ms"]), nudgesMs.contains(ms) else { return nil }
      return .nudge(ms: Int(ms))
    case "feel":
      let sense = msg["sense"] as? String
      return sense == "early" || sense == "late" ? .feel(early: sense == "early") : nil
    default:
      return nil
    }
  }
}

/// The TV's open end of a room. Listens until stopped, reconnecting through the
/// relay's 4-minute stream limit and network drops.
public final class RelayLink: @unchecked Sendable {
  public let room: String
  private let url: URL
  private var task: Task<Void, Never>?

  public init(room: String, base: URL = Companion.base) {
    self.room = room
    url = Companion.relayURL(room: room, base: base)
  }

  public func start(onCommand: @escaping @MainActor (PhoneCommand) -> Void) {
    task?.cancel()
    let url = url
    task = Task.detached {
      while !Task.isCancelled {
        do {
          var req = URLRequest(url: url, timeoutInterval: 60) // idle timeout; the relay pings every 20s
          req.setValue("text/event-stream", forHTTPHeaderField: "Accept")
          let (bytes, _) = try await URLSession.shared.bytes(for: req)
          for try await line in bytes.lines where line.hasPrefix("data:") {
            let payload = line.dropFirst(5).drop(while: { $0 == " " })
            if let cmd = Companion.parseCommand(Data(payload.utf8)) { await onCommand(cmd) }
          }
        } catch {
          try? await Task.sleep(nanoseconds: 2_000_000_000) // offline — back off before retrying
        }
      }
    }
  }

  public func send<T: Encodable>(_ message: T) {
    guard let body = try? JSONEncoder().encode(message) else { return }
    var req = URLRequest(url: url)
    req.httpMethod = "POST"
    req.setValue("text/plain", forHTTPHeaderField: "Content-Type")
    req.httpBody = body
    URLSession.shared.dataTask(with: req).resume()
  }

  public func stop() {
    task?.cancel()
  }
}
